docs/proposals/compute-resource-metrics-api.md

```diff
@@ -0,0 +1,195 @@
+  to fetch its node and pods metrics. Then Heapster will have to implement the
+  APIs mentioned above.
+
+Our initial plan is to follow the push based system.
+
+The API server will not hold any of this data directly. For our initial
+purposes, it will hold the most recent metrics obtained from each node in
+etcd. Then, when polled for metrics, the API server would only serve said most
+recent data per node.
+
+If a pod gets deleted, the API server will get rid of any metrics it may
+currently be holding for it.
+
+The clients watching the metrics data may cache it for longer periods of time.
+The clearest example would be Heapster.
+
+In the future, we might want to store the metrics differently:
+
+* via heapster - Since heapster keeps data for a period of time, we could
+  redirect requests to the API server to heapster instead of using etcd. This
+  would also allow serving metrics other than the latest ones.
+
+An edge case that this proposal doesn't take into account is kubelets being
+restarted. If any of them are, with a simple implementation they would lose
+historical data and thus take hours to gather enough information to provide
+relevant metrics again. We might want to use persistent storage directly or in
+the future to improve that situation.
+
+## Data structure
+
+```Go
+type DerivedPodMetrics struct {
+	TypeMeta
+	ObjectMeta // should have pod name
+	// the key is the container name
+	Containers map[string]ResouresMetrics
+}
+
+type DerivedNodeMetrics struct {
+	TypeMeta
+	ObjectMeta // should have node name
+	NodeMetrics      ResourcesMetrics
+	SystemContainers map[string]ResourcesMetrics
+}
+
+type ResourcesMetrics map[resource.Type]DerivedMetrics
+
+type DerivedMetrics struct {
+	// Time at which these metrics were gathered
+	Time util.Time `json:"timestamp"`
+
+	// 1m, 1h and 1d as a start
+	Metrics map[time.Duration]Metrics
+}
+
+type Metrics struct {
+	// Average over the collected sample.
+	Mean resource.Quantity `json:"mean"`
+	// Max seen over the collected sample.
+	Max resource.Quantity `json:"max"`
+	// 95th percentile over the collected sample.
+	NinetyFive resource.Quantity `json:"95th"`
+}
+
+type RawPodMetrics struct {
+	TypeMeta
+	ObjectMeta // should have pod name
+	// the key is the container name
+	Containers map[string]struct {
+		ContainerReference *Container
+		RawMetrics         []RawMetric
+	}
+}
+
+type RawNodeMetrics struct {
+	TypeMeta
+	ObjectMeta // should have node name
+	NodeMetrics      []RawMetric
+	SystemContainers map[string][]RawMetric
+}
+
+type RawMetric {
+	Time util.Time
+	CPU  CpuStats
+	Mem  MemoryStats
+}
+
+// All CPU usage metrics are cumulative from the creation of the container
+type CpuStats struct {
+	Usage CpuUsage `json:"usage"`
+	// Smoothed average of number of runnable threads x 1000.
+	// We multiply by thousand to avoid using floats, but preserving precision.
+	// Load is smoothed over the last 10 seconds. Instantaneous value can be read
+	// from LoadStats.NrRunning.
```

Comment:
I think this is going to be much noisier than just using the delta of (cpuact.usage) over some time window.